### PR TITLE
WMAP improvement

### PR DESCRIPTION
**Pull request by flibustier**
- New output for `wmap_sites -s` 
- New option in wmap_sites for unicode tree output (`wmap_sites -s [ids] (level) (unicode output true/false)`)
- Filtering 404 in the tree output

You can see some of the informations in issue [#8089](https://github.com/rapid7/metasploit-framework/issues/8089)

## Verification

[asciicast](https://asciinema.org/a/a1w5c3dh87ecf7pxsicbg8fbn?speed=3)

### CODE REVIEW

File: plugins/wmap.rb [hunk 1]
```diff
@@ -1552,69 +1554,106 @@ def view_site_tree(urlstr, md, ld)
...
-        end
+        load_tree_core(req, wtree)
       end
 
-      return wtree
+      wtree
     end
 
-    #
-    # Print Tree structure. Still ugly
```
Comment (wvu-r7) [inline plugins/wmap.rb]: Quite!

File: plugins/wmap.rb [hunk 2]
```diff
@@ -147,23 +147,28 @@ def cmd_wmap_sites(*args)
...
 
-          if not u
-            return
-          end
+          return unless u
 
           if l == nil or l.empty?
             l = 200
-            s = true
+            o = 'true'
```
Comment (wvu-r7) [inline plugins/wmap.rb]: Why is this not Boolean `true`?

File: plugins/wmap.rb [hunk 3]
```diff
@@ -147,23 +147,28 @@ def cmd_wmap_sites(*args)
...
           else
-            l = l.to_i
-            s = false
+            # Add check if unicode parameters is the second one
+            if l == 'true' or l == 'false'
```
Comment (wvu-r7) [inline plugins/wmap.rb]: `||` is preferred over `or`.

File: plugins/wmap.rb [hunk 4]
```diff
@@ -147,23 +147,28 @@ def cmd_wmap_sites(*args)
         when '-s'
           u = args.shift
           l = args.shift
-          s = args.shift
+          o = args.shift
```
Comment (wvu-r7) [inline plugins/wmap.rb]: Why the var change?

File: plugins/wmap.rb (hunk 2 above)
Comment (wvu-r7) [inline plugins/wmap.rb]: Ah, it comes from the arg. You should be resolving the string to a Boolean.

File: plugins/wmap.rb (hunk 4 above)
Comment (wvu-r7) [inline plugins/wmap.rb]: I assume for "output?"

File: plugins/wmap.rb [hunk 5]
```diff
@@ -147,23 +147,28 @@ def cmd_wmap_sites(*args)
...
+            # Add check if unicode parameters is the second one
+            if l == 'true' or l == 'false'
+              o = l
+              l = 200
+            else
+              l = l.to_i
+            end
           end
 
+          o = (o == 'true')
```
Comment (wvu-r7) [inline plugins/wmap.rb]: Looks like you're resolving to Boolean down here. Can you move it earlier?

**Comment by wvu-r7** [comment]
This diff is difficult to read due to the gaps. Apologies if the comments don't read linearly. It seems you've already accounted for all my review points. Nice work!

File: plugins/wmap.rb (hunk 5 above)
Comment (wvu-r7) [inline plugins/wmap.rb]: Hmm, maybe not.

**Comment by wvu-r7** [comment]
So, do you think you can add Unicode detection like `tree`?

**Comment by busterb** [comment]
@wvu-r7 I believe our current approach is to assume everything but Windows is Unicode, or at least that's how the Meterpreter console does it.

        OptBool.new('EnableUnicodeEncoding', [true, "Automatically encode UTF-8 strings as hexadecimal", Rex::Compat.is_windows]),

Another approach would be to check the LANG environment variable contains UTF-8

**Comment by busterb** [comment]
If you come up with a common method we can share between the two locations, I'll bet there are some other areas we could start fixing this too. Maybe we should add Rex::Compat.output_unicode or something.

File: plugins/wmap.rb (hunk 4 above)
Comment (J0hn-) [inline plugins/wmap.rb]: Yes exactly ! 's' was useless before so I recycle it !

File: plugins/wmap.rb (hunk 3 above)
Comment (J0hn-) [inline plugins/wmap.rb]: Sorry ! (Newbie in Ruby)

File: plugins/wmap.rb (hunk 5 above)
Comment (J0hn-) [inline plugins/wmap.rb]: That's right, an other option is this :
```
          o = args.shift
          o = (o == 'true')

          return unless u

          if l == nil or l.empty?
            l = 200
            o = true
          else
            # Add check if unicode parameters is the second one
            if l == 'true' || l == 'false'
              o = (l == 'true')
              l = 200
            else
              l = l.to_i
            end
          end
```
But it's involves two resolutions, not necessarily prettier…

File: plugins/wmap.rb (hunk 3 above)
Comment (wvu-r7) [inline plugins/wmap.rb]: Lol, this was the only valid criticism. Nice work!

**Comment by wvu-r7** [comment]
Thanks, great work!

**Comment by J0hn-** [comment]
Thanks !! :)

**Comment by wvu-r7** [comment]
# Release Notes

This improves the WMAP plugin by drawing a prettier Unicode tree for the `wmap_sites -s` command.

**Pull request merged**